### DISCUSSION TITLE
Formatting C source with emscripten's EM_ASM

### DISCUSSION THREAD

**Discussion by jonahgraham** [prettier/prettier#16504] [Q&A]
Does anyone have advice on how to deal with prettier misformatting C files that contain EM_ASM macros?

For example, this is my source:

```c
#include <emscripten.h>

void formatting(void)
{
    EM_ASM(
        methodWithCallback({
            callback: function() {
                return 1;
            }
        });
    );
}
```

But with prettier I end up with:

```c
#include <emscripten.h>

void formatting(void)
{
    EM_ASM(
        methodWithCallback({
            callback: function() {
                return 1;
}
});
    );
    }
```

For the purposes of reproducing without emscripten the `#include` can be replaced with `#define EM_ASM(...)` for the same effect:

```c
#define EM_ASM(...)

void formatting(void)
{
    EM_ASM(
        methodWithCallback({
            callback: function() {
                return 1;
            }
        });
    );
}
```

---

The stretch goal would be to get prettier to format the code inside the EM_ASM using javascript formatter too.

Any advice? Is the first part a bug, or outside the scope of prettier?

**Comment by Labeeb2339** [comment]
The formatter is seeing the `EM_ASM` invocation as C syntax. The JavaScript inside the macro is not a separate JavaScript AST, so a C parser/printer cannot safely determine its braces, indentation, or where the macro argument ends. Formatting the nested JavaScript automatically would require a language-aware plugin that extracts the macro payload, runs the JavaScript parser, and prints it back with correct source mapping.

For the current setup, the safe options are to exclude macro-containing files from the C formatter, or use the formatter's ignore marker immediately before the `EM_ASM` node if the C plugin you use supports `prettier-ignore` comments. Format the JavaScript source separately before embedding it. Avoid relying on a C formatter to infer that arbitrary macro arguments contain JavaScript, because similar macros can legally contain strings or other token sequences.

**Reply by jonahgraham** [reply]
Would prettier be able to be upgraded to understand EM_ASM in C code, in the same way the vue formatter supports multiple embedded languages? Or are you sharing this is outside of the scope and ability of Prettier?
